train_segmentation.py

```diff
@@ -78,40 +102,81 @@ def create_dataloader(samples_folder, batch_size, num_devices, params):
+                                                                             dataset=subset,
+                                                                             aug_type='geometric',
+                                                                             dontcare=dontcare_val,
+                                                                             crop_size=crop_size),
+                                       totensor_transform=aug.compose_transforms(params=params,
+                                                                                 dataset=subset,
+                                                                                 input_space=BGR_to_RGB,
+                                                                                 scale=scale,
+                                                                                 dontcare2backgr=dontcare2backgr,
+                                                                                 dontcare=dontcare_val,
+                                                                                 aug_type='totensor'),
                                        params=params,
                                        debug=debug))
     trn_dataset, val_dataset, tst_dataset = datasets
 
     # https://discuss.pytorch.org/t/guidelines-for-assigning-num-workers-to-dataloader/813/5
-    num_workers = num_devices * 4 if num_devices > 1 else 4
+    num_workers = len(gpu_devices_dict.keys()) * 4 if len(gpu_devices_dict.keys()) > 1 else 4
 
     samples_weight = torch.from_numpy(samples_weight)
     sampler = torch.utils.data.sampler.WeightedRandomSampler(samples_weight.type('torch.DoubleTensor'),
                                                              len(samples_weight))
+    eval_batch_size = batch_size
+
+    if gpu_devices_dict and calc_eval_bs:
+        max_pix_per_mb_gpu = 280  # TODO: this value may need to be finetuned
+        calc_eval_batchsize(gpu_devices_dict, batch_size, sample_size, max_pix_per_mb_gpu)
 
     trn_dataloader = DataLoader(trn_dataset, batch_size=batch_size, num_workers=num_workers, sampler=sampler,
                                 drop_last=True)
-    val_dataloader = DataLoader(val_dataset, batch_size=1, num_workers=num_workers, shuffle=False,
+    val_dataloader = DataLoader(val_dataset, batch_size=eval_batch_size, num_workers=num_workers, shuffle=False,
                                 drop_last=True)
-    tst_dataloader = DataLoader(tst_dataset, batch_size=1, num_workers=num_workers, shuffle=False,
+    tst_dataloader = DataLoader(tst_dataset, batch_size=eval_batch_size, num_workers=num_workers, shuffle=False,
                                 drop_last=True) if num_samples['tst'] > 0 else None
 
     return trn_dataloader, val_dataloader, tst_dataloader
 
 
-def get_num_samples(samples_path, params):
+def calc_eval_batchsize(gpu_devices_dict: dict, batch_size: int, sample_size: int, max_pix_per_mb_gpu: int = 280):
```

Comment:
Again, this may not reach consensus. Trying to automatically calculate evaluation batch size based on an approximate "maximum pixels per Mb of GPU" (this also depends on models. I overestimated the value to make sure it passes even with large models such as deeplabv3)